src/wui/soldierlist.cc

```diff
@@ -481,22 +506,28 @@ SoldierList::SoldierList(UI::Panel& parent,
 	UI::Box* buttons = new UI::Box(this, UI::PanelStyle::kWui, 0, 0, UI::Box::Horizontal);
 
 	bool can_act = ibase_.can_act(building_or_ship_.owner().player_number());
-	if (building_or_ship.descr().type() == Widelands::MapObjectType::MILITARYSITE) {
-		upcast(Widelands::MilitarySite, ms, &building_or_ship);
+	{
+		// Make sure the creation order is consistent with enum SoldierPreference!
 		soldier_preference_.add_button(buttons, UI::PanelStyle::kWui, Vector2i::zero(),
 		                               g_image_cache->get("images/wui/buildings/prefer_rookies.png"),
 		                               _("Prefer rookies"));
-		soldier_preference_.add_button(buttons, UI::PanelStyle::kWui, Vector2i(32, 0),
+		soldier_preference_.add_button(buttons, UI::PanelStyle::kWui, Vector2i::zero(),
 		                               g_image_cache->get("images/wui/buildings/prefer_heroes.png"),
 		                               _("Prefer heroes"));
+		soldier_preference_.add_button(buttons, UI::PanelStyle::kWui, Vector2i::zero(),
+		                               g_image_cache->get("images/wui/buildings/prefer_any.png"),
+		                               _("No preference"));
 		UI::Radiobutton* button = soldier_preference_.get_first_button();
 		while (button != nullptr) {
 			buttons->add(button);
 			button = button->next_button();
 		}
 
 		soldier_preference_.set_state(
-		   ms->get_soldier_preference() == Widelands::SoldierPreference::kHeroes ? 1 : 0, false);
+		   static_cast<uint8_t>(ms != nullptr ? ms->get_soldier_preference() :
+		                        wh != nullptr ? wh->get_soldier_preference() :
+                                              ship->get_soldier_preference()),
```

Comment:
So it can be trainingsite as well, so instead of nesting one more ternary op, maybe this should be done right after the upcasts in lines 476-478. I don't think you use these upcasted objects for anything else, so you don't even have to try all of them after the first successful one.